python/lsst/pipe/base/connections.py

```diff
@@ -456,46 +457,124 @@ def buildDatasetRefs(self, quantum: Quantum) -> typing.Tuple[InputQuantizedConne
+            of its expected inputs do not exist, and if possible, should be
+            pruned from the QuantumGraph.
         Exception
-            Overrides of this function have the option of raising an Exception
-            if a field in the input does not satisfy a need for a corresponding
-            pipelineTask, i.e. no reference catalogs are found.
+            Overrides of this function may raise other exceptions if a field in
+            the input does not satisfy a need for a corresponding pipelineTask,
+            i.e. no reference catalogs are found, and this should be considered
+            a pre-execution error, not a no-work condition.
+
+        Notes
+        -----
+        The base class implementation performs useful checks and should be
+        called via `super` by most custom implementations.  It always returns
+        an empty iterable, because it makes no changes.
+
+        Implementations are permitted (but not required) to be implemented
+        as generators, and raise the documented exceptions during iteration
+        rather than immediately upon return.
         """
-        for connection in itertools.chain(iterConnections(self, "inputs"),
-                                          iterConnections(self, "prerequisiteInputs")):
-            refs = datasetRefMap[connection.name]
+        for name, connection, refs in inputs:
             if not connection.multiple and len(refs) > 1:
                 raise ScalarError(
                     f"Found multiple datasets {', '.join(str(r.dataId) for r in refs)} "
-                    f"for scalar connection {connection.name} ({refs[0].datasetType.name})."
+                    f"for scalar connection {label}.{name} ({connection.name}) "
+                    f"for quantum data ID {dataId}."
+                )
+            if not connection.optional and not refs:
+                if isinstance(connection, PrerequisiteInput):
+                    # This branch should only be possible during QG generation,
+                    # or if someone deleted the dataset between making the QG
+                    # and trying to run it.  Either one should be a hard error.
+                    raise FileNotFoundError(
+                        f"No datasets found for non-optional connection {label}.{name} ({connection.name}) "
+                        f"for quantum data ID {dataId}."
+                    )
+                else:
+                    # This branch should be impossible during QG generation,
+                    # because that algorithm can only make quanta whose inputs
+                    # are either already present or should be created during
+                    # execution.  It can trigger during execution if the input
+                    # wasn't actually created by an upstream task in the same
+                    # graph.
+                    raise NoWorkFound(label, name, connection)
+        yield from ()
```

Comment:
As per a review you did for me, do you want this to just be `return iter(())`

Reply:
Yeah, I probably do; reduces the chance that someone delegates to super just for exception-raising checks and nothign happens because generator.  But I'm probably going to make it mapping anyway, as per previous comment.